### PR TITLE
refactor: getLatestMessageDisplayText

### PR DESCRIPTION
**Pull request by Enigma-I-am**
## 🎯 Goal
The goal of this PR is to refactor and test the getLatestMessageDisplayText function in the `useLatestMessagePreview` file.

## 🛠 Implementation details
Refactored this method by moving some of it's verbose parts to functions of their own and also wrote a test for the hook

## 🎨 UI Changes

<details>
<summary>iOS</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

<details>
<summary>Android</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

## ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Commits follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) spec
- [x] New code is covered by tests
- [ ] Screenshots added for visual changes
- [ ] Documentation is updated

### CODE REVIEW

File: package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts [hunk 1]
```diff
@@ -29,6 +29,43 @@ export type LatestMessagePreview<
   status: number;
 };
 
+const messageOwner = <
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts:32]: What does this function do? Let's give it a better name!

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    });
+  });
+
+  it('should return latest attachment', async () => {
+    const { result } = renderHook(
+      () =>
+        useLatestMessagePreview(CHANNEL_WITH_MESSAGES_ATTACHMENTS, FORCE_UPDATE, LATEST_MESSAGE),
+      { wrapper: ChatProvider },
+    );
+    console.log(result.current.created_at);
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:312]: please remove console.log

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+      });
+    });
+  });
+
+  it('should return latest command', async () => {
+    const { result } = renderHook(
+      () => useLatestMessagePreview(CHANNEL_WITH_MESSAGES_COMMAND, FORCE_UPDATE, LATEST_MESSAGE),
+      { wrapper: ChatProvider },
+    );
+    console.log(result.current.created_at);
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:277]: please remove console.log

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+      });
+    });
+  });
+
+  it('should return latest text', async () => {
+    const { result } = renderHook(
+      () => useLatestMessagePreview(CHANNEL_WITH_MESSAGES_TEXT, FORCE_UPDATE, LATEST_MESSAGE),
+      { wrapper: ChatProvider },
+    );
+    console.log(result.current.created_at);
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:241]: please remove console.log

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+      expect(result.current).toBeTruthy();
+    });
+  });
+
+  it('should return a channel', async () => {
+    const { result } = renderHook(
+      () => useLatestMessagePreview(channelWithMessages, FORCE_UPDATE, LATEST_MESSAGE),
+      { wrapper: ChatProvider },
+    );
+    console.log(result.current);
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:215]: please remove console.log

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+  const clientUser = generateUser();
+  let chatClient: StreamChat<DefaultGenerics> | StreamChat<DefaultStreamChatGenerics>;
+
+  const channelName = 'okechukwu';
+  const channel = {
+    data: { name: channelName },
+    state: { messages: [] },
+  } as unknown as Channel<DefaultStreamChatGenerics>;
+
+  const CHANNEL_WITH_MESSAGES_TEXT = {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:37]: Please move all costs to a separate mocks file ;)

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 7]
```diff
@@ -0,0 +1,360 @@
+import React, { FC } from 'react';
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:1]: > SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=GetStream_stream-chat-react-native&pullRequest=1425)
> 
> [Bug](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_coverage&view=list) [54.3% Coverage](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_duplicated_lines_density&view=list)

Please increase code coverage to pass quality gate (%80)

**Review by vanGalilea — CHANGES_REQUESTED** [review]

**Comment by Enigma-I-am** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=GetStream_stream-chat-react-native&pullRequest=1425)
> 
> [Bug](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=GetStream_stream-chat-react-native&pullRequest=1425&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_coverage&view=list) [94.4% Coverage](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=GetStream_stream-chat-react-native&pullRequest=1425&metric=new_duplicated_lines_density&view=list)

@vanGalilea test coverage is now above 80%

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        messageObject: {
+          type: 'deleted',
+        },
+        previews: [{ bold: false, text: 'Message deleted' }],
+        status: 0,
+      });
+    });
+  });
+
+  it('should return message deleted', async () => {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:117]: This test title is incorrect right?

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    await waitFor(() => {
+      expect(result.current).toEqual({
+        created_at: 'LT',
+        previews: [{ bold: false, text: 'Nothing yet...' }],
+        status: 0,
+      });
+    });
+  });
+
+  it('should return latest text', async () => {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:131]: The title of the test should be more specific, we're expecting the whole result is there.

But at the end of the day, we care about what is being rendered to the user.

Comment (vanGalilea) [reply package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:131]: we say "should return latest text" if that's what you want to test, just expect `result.current.reviews` (or whatever is being rendered)

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+        previews: [
+          { bold: false, text: '' },
+          { bold: false, text: 'Empty message...' },
+        ],
+        status: 0,
+      });
+    });
+  });
+
+  it('should return a channel with a mentioned user', async () => {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:198]: same for the name here, do we return a channel here or actually a message preview with a mentioned user?

Comment (vanGalilea) [reply package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:198]: Please be more spec. with assertion

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+          created_at: new Date('2021-02-12T12:12:35.862Z'),
+          deleted_at: new Date('2021-02-12T12:12:35.862Z'),
+          id: 'ljkblk',
+          user: {
+            id: 'okechukwu',
+          },
+        },
+        previews: [
+          { bold: true, text: '@okechukwu: ' },
+          { bold: false, text: '🏙 Attachment...' },
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:311]: That's what the user sees so let's only assert that, what are your thoughts on that?

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        previews: [
+          { bold: true, text: '@okechukwu: ' },
+          { bold: false, text: 'jkbkbiubicbi' },
+        ],
+        status: 0,
+      });
+    });
+  });
+
+  it('should return a channel with an empty message', async () => {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:166]: again not a channel here, let's give a test a better and realistic name

File: package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+          created_at: new Date('2021-02-12T12:12:35.862Z'),
+          deleted_at: new Date('2021-02-12T12:12:35.862Z'),
+          id: 'jbkjb',
+          user: {
+            id: 'okechukwu',
+          },
+        },
+        previews: [
+          { bold: true, text: '@okechukwu: ' },
+          { bold: false, text: '/giphy' },
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/__tests__/useLatestMessagePreview.test.tsx:257]: let's only assert what the users sees etc

File: package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts [hunk 14]
```diff
@@ -29,6 +29,43 @@ export type LatestMessagePreview<
...
+  return '';
+};
+
+const getMentionUsers = <
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
+>(
+  mentionedUser: UserResponse<StreamChatGenerics>[] | undefined,
+) => {
+  if (Array.isArray(mentionedUser)) {
+    const mentionUSerString = mentionedUser.reduce((acc, cur) => {
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts:57]:
```suggestion
    const mentionUserString = mentionedUser.reduce((acc, cur) => {
```

File: package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts [hunk 15]
```diff
@@ -29,6 +29,43 @@ export type LatestMessagePreview<
...
+      const userName = cur.name || cur.id || '';
+      if (userName) {
+        acc += `${acc.length ? '|' : ''}@${userName}`;
+      }
+      return acc;
+    }, '');
+    return mentionUSerString;
```
Comment (vanGalilea) [inline package/src/components/ChannelPreview/hooks/useLatestMessagePreview.ts:64]:
```suggestion
    return mentionUserString;
```

**Review by vanGalilea — CHANGES_REQUESTED** [review]

**Comment by Enigma-I-am** [comment]
@vanGalilea updates have been puushed

**Review by vanGalilea — APPROVED** [review]

**Pull request merged**